NH-119712 Discovery Collector to not require OpenTelemetry Operator

jakub-racek-swi opened the pull request:
jira: https://swicloud.atlassian.net/browse/NH-119712

Summary:
- `targetAllocator` templates moved to their own folder, no subcharts required
- updated values.yaml with an entry for the `targetAllocator` version
- `discovery_collector` (OpenTelemetryCollector resource) transformed into a statefulset resource with a config map
- added generated unit tests for `targetAllocator`

Verified on docker desktop and e2e cluster

gantrior commented on deploy/helm/templates/targetAllocator/target-allocator-config-map.yaml:
```diff
@@ -0,0 +1,64 @@
+    allocation_strategy: consistent-hashing
+    collector_namespace: {{ .Release.Namespace }}
+    collector_selector:
+      matchlabels:
```
@jakub-racek-swi isn't this actual mistake? `collector_selector` reference LabelSelector from kubernetes API - https://github.com/open-telemetry/opentelemetry-operator/blob/776ce57330fd42eb682d50d4e76f28ebeed3413e/cmd/otel-allocator/internal/config/config.go#L67

pstranak-sw commented on deploy/helm/templates/targetAllocator/target-allocator-config-map.yaml:
```diff
@@ -0,0 +1,64 @@
+    collector_selector:
+      matchLabels:
+        app.kubernetes.io/component: discovery-collector
+        app.kubernetes.io/instance: {{ .Release.Name }}
```
What deployment/service/... is this supposed to match?

jakub-racek-swi commented on deploy/helm/values.yaml:
```diff
@@ -101,6 +101,11 @@ otel:
         targetAllocator:
+          image:
+            repository: ghcr.io/open-telemetry/opentelemetry-operator/target-allocator
+            tag: "0.116.0"
```
yes, we should probably match the operator version

jakub-racek-swi commented on deploy/helm/templates/operator/discovery-collector.yaml:
```diff
@@ -1,349 +1,175 @@
-        metrics:
-{{ toYaml .Values.otel.metrics.filter | indent 10 }}
+  serviceName: {{ include "common.fullname" (tuple . "-discovery-collector") }}
+  replicas: 1
```
looks like I forgot about this 😄 
I don't like how we don't check the structure at all in the chart right now, so maybe I'll make it more strict

jakub-racek-swi commented:
@pstranak-sw to recap:
Added annotations where they were missing
Added HPA - the changes to values should be safe, I don't think we'll need to cover more of the options, seems to be edge cases.

jakub-racek-swi commented on deploy/helm/templates/operator/discovery-collector.yaml:
```diff
@@ -1,349 +1,175 @@
 metadata:
-  # name is limitted to 41, as `-collector` is autoadded by the operator and statefulset can have maximum 52 characters
-  name: {{ include "common.fullname" (tuple . "-discovery" 41) }} 
+  name: {{ include "common.fullname" (tuple . "-discovery-collector" 51) }}
```
StatefulSet names have a maximum of 52 characters, otherwise they'll fail. Not sure why we only use 51 of them, but I'm assuming the person who originally wrote that had a good reason
